File: Data Structures & Algorithms/min-cost-to-connect-points/submission-4.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)

        adj = defaultdict(list)
        visited = set()

        # build edges
        for index, point in enumerate(points):
            x, y = point
            for j in range(index + 1, n):
                nx, ny = points[j]
                dist = abs(x - nx) + abs(y - ny)
                adj[index].append((dist, j))
                adj[j].append((dist, index))

        totalCost = 0
        frontier = [[0, 0]]
        while len(visited) < n:

            cost, index = heapq.heappop(frontier)
            if index in visited:
                continue
            totalCost += cost
            visited.add(index)
            for neighborCost, neighbor in adj[index]:
                if neighbor not in visited:
                    heapq.heappush(frontier, (neighborCost, neighbor))

        return totalCost

File: Data Structures & Algorithms/climbing-stairs/submission-4.cpp
class Solution {
public:
    int climbStairs(int n) {
        vector<int> cache(n + 1, 0);

        cache[0] = 1;
        cache[1] = 2;
        if (n < 2)
        {
            return cache[n-1];
        }

        for (int index = 3; index < n + 1; index++)
        {
            cache[index - 1] = cache[index - 2] + cache[index - 3];
        }

        return cache[n - 1];
    }
};
